alphadia/workflow/peptidecentric.py

```diff
@@ -444,89 +405,205 @@ def calibration(self):
+            [optimizer]
+            for optimizer in optimizers
+            if isinstance(optimizer, searchoptimization.AutomaticOptimizer)
+        ]
+
+        order_of_optimization = (
+            targeted_optimizers + automatic_optimizers
+            if any(targeted_optimizers)
+            else automatic_optimizers
+        )
+
+        self.reporter.log_string(
+            "Starting initial classifier training and precursor identification.",
+            verbosity="progress",
+        )
+
+        self.eg_idxes_for_calibration, precursor_df, fragments_df = (
+            self.extract_optimization_data(
+                self.config["calibration"]["min_precursors_for_optimization"],
+                self.config["calibration"]["min_training_iterations"],
+            )
+        )
+
+        self.com.fit({"classifier_version": self.fdr_manager.current_version})
+
+        precursor_df_filtered, fragments_df_filtered = self.filter_dfs(
+            precursor_df, fragments_df
+        )
 
-            self.start_of_epoch(current_epoch)
+        self.recalibration(precursor_df_filtered, fragments_df_filtered)
 
-            features = []
-            fragments = []
-            for current_step, (start_index, stop_index) in enumerate(self.batch_plan):
-                self.start_of_step(current_step, start_index, stop_index)
+        self.reporter.log_string(
+            "Required number of precursors found and required number of training iterations performed. Starting search parameter optimization.",
+            verbosity="progress",
+        )
 
-                eg_idxes = self.elution_group_order[start_index:stop_index]
+        self.reporter.log_string(
+            "=== Performing initial optimization on extracted data. ===",
+            verbosity="info",
+        )
+        # Perform an initial optimization step based on the extracted data to update the initial search parameters.
+        # This ensures that the classifier is trained at least once prior to the end of optimization, even if the min_steps parameter is 0.
+        for optimizers in order_of_optimization:
+            for optimizer in optimizers:
+                optimizer.step(precursor_df_filtered, fragments_df_filtered)
+
+        for optimizers in order_of_optimization:
+            for current_step in range(self.config["calibration"]["max_steps"]):
+                if np.all([optimizer.has_converged for optimizer in optimizers]):
+                    self.reporter.log_string(
+                        f"Optimization finished for {', '.join([optimizer.parameter_name for optimizer in optimizers])}.",
+                        verbosity="progress",
+                    )
+
+                    for optimizer in optimizers:
+                        if isinstance(optimizer, searchoptimization.AutomaticOptimizer):
+                            optimizer.plot()
+
+                    break
                 batch_df = self.spectral_library._precursor_df[
```

Comment:
do we need to regenerate this every iteration? isn't this frozen for the whole optimization?

Reply:
okay... let's keep this for now but i guess there is some future optimization potential